packages/core/src/stateUtils.ts

```diff
@@ -523,76 +510,28 @@ function isHistoryNode<TContext, TEvent extends EventObject>(
   return stateNode.type === 'history';
 }
 
-/**
- * Retrieves state nodes from a relative path to the state node.
- *
- * @param relativePath The relative path from the state node
- */
-function getFromRelativePath<TContext, TEvent extends EventObject>(
-  stateNode: StateNode<TContext, any, TEvent>,
-  relativePath: string[]
-): Array<StateNode<TContext, any, TEvent>> {
-  if (!relativePath.length) {
-    return [stateNode];
-  }
-  const [stateKey, ...childStatePath] = relativePath;
-  if (!stateNode.states) {
-    throw new Error(
-      `Cannot retrieve subPath '${stateKey}' from node with no states`
-    );
-  }
-  const childStateNode = getStateNode(stateNode, stateKey);
-  if (isHistoryNode(childStateNode)) {
-    return resolveHistory(childStateNode);
-  }
-  if (!stateNode.states[stateKey]) {
-    throw new Error(
-      `Child state '${stateKey}' does not exist on '${stateNode.id}'`
-    );
-  }
-  return getFromRelativePath(stateNode.states[stateKey], childStatePath);
-}
-/**
- * Returns the leaf nodes from a state path relative to the state node.
- *
- * @param relativeStateNode The relative state node to retrieve the state nodes
- * @param state The previous state to retrieve history
- * @param resolveInitialStateNodes Whether state nodes should resolve to initial child state nodes
- */
-export function getRelativeStateNodes<TContext, TEvent extends EventObject>(
-  relativeStateNode: StateNode<TContext, any, TEvent>,
-  state?: State<TContext, TEvent>,
-  resolveInitialStateNodes: boolean = true
-): Array<StateNode<TContext, any, TEvent>> {
-  return resolveInitialStateNodes
-    ? isHistoryNode(relativeStateNode)
-      ? resolveHistory(relativeStateNode, state)
-      : getInitialStateNodes(relativeStateNode)
-    : [relativeStateNode];
-}
-
 export function getInitialStateNodes<TContext, TEvent extends EventObject>(
   stateNode: StateNode<TContext, any, TEvent>
 ): Array<StateNode<TContext, any, TEvent>> {
-  if (isLeafNode(stateNode)) {
-    return [stateNode];
-  }
-  // Case when state node is compound but no initial state is defined
-  if (stateNode.type === 'compound' && !stateNode.initial) {
-    if (!IS_PRODUCTION) {
-      warn(
-        false,
-        `Compound state node '${stateNode.id}' has no initial state.`
-      );
+  const transitions = [
+    {
+      target: [stateNode],
+      source: stateNode,
+      actions: [],
+      eventType: 'init'
     }
-    return [stateNode];
-  }
-  const initialStateNodePaths = toStatePaths(stateNode.initialStateValue!);
-  return flatten(
-    initialStateNodePaths.map(initialPath =>
-      getFromRelativePath(stateNode, initialPath)
-    )
+  ];
+  const mutStatesToEnter = new Set<StateNode<TContext, any, TEvent>>();
+  const mutStatesForDefaultEntry = new Set<StateNode<TContext, any, TEvent>>();
+
+  computeEntrySet(
```

Comment:
this is quite hacky but solves the problem of resolving initial states recursively with the existing algorithm

 i would like to refactor this in the future, but for the time being it seems OK - @davidkpiano's work on representing initial states as transitions will affect this anyway, so that would be a good time to refactor this